pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java

```diff
@@ -1241,9 +1322,71 @@ public DatabaseMetaData getMetaData() throws SQLException {
     return metadata;
   }
 
+  /**
+   * @see  #setTypeMap(java.util.Map)
+   */
+  private static Set<Class<?>> getAllClasses(Class<?> clazz) {
+    Set<Class<?>> classes = new HashSet<Class<?>>();
+    Class<?> current = clazz;
+    do {
+      classes.add(current);
+      for (Class<?> iface : clazz.getInterfaces()) {
```

Comment:
what about interface hierarchy? For example, ArrayList implements List which extends Collection. I believe that your set would include ArrayList and List, but not Collection. (Actually it would, but because ArrayList ends up extending AbstractCollection).

Reply:
Good point.  Thank you.  I'll work on it.

Reply:
I have reworked this to a recursive implementation to get all interfaces, and extended interfaces, all the way up.  It also adds to the inferenceMap directly, instead of building the intermediate set of classes.

https://github.com/pgjdbc/pgjdbc/pull/1378/commits/3249827059d113bcdd9694a7213cc77809e0cbe3

Reply:
I have added a test case for the scenario you describe, with a rather contrived interface and class hierarchy: https://github.com/pgjdbc/pgjdbc/pull/1378/commits/72597e4b262cb16284350b89993ccb2fd2ab44f4